planner/cascades/transformation_rules.go

```diff
@@ -2502,3 +2505,87 @@ func (r *PullSelectionUpApply) OnTransform(old *memo.ExprIter) (newExprs []*memo
 	newApplyGroupExpr.SetChildren(outerChildGroup, old.Children[1].GetExpr().Children[0])
 	return []*memo.GroupExpr{newApplyGroupExpr}, false, false, nil
 }
+
+// MergeAdjacentWindow merge adjacent Window.
+type MergeAdjacentWindow struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentWindow creates a new Transformation MergeAdjacentWindow.
+// The pattern of this rule is `Window -> Window`.
+func NewRuleMergeAdjacentWindow() Transformation {
+	rule := &MergeAdjacentWindow{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandWindow,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandWindow, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentWindow) Match(expr *memo.ExprIter) bool {
+	curWinPlan := expr.GetExpr().ExprNode.(*plannercore.LogicalWindow)
+	nextGroupExpr := expr.Children[0].GetExpr()
+	nextWinPlan := nextGroupExpr.ExprNode.(*plannercore.LogicalWindow)
+	nextSchema := nextGroupExpr.Group.Prop.Schema
+	nextGroupChildren := nextGroupExpr.Children
+	ctx := expr.GetExpr().ExprNode.SCtx()
+
+	// Whether Partition, OrderBy and Frame parts are the same.
+	if !(curWinPlan.EqualPartitionBy(ctx, nextWinPlan) &&
+		curWinPlan.EqualOrderBy(ctx, nextWinPlan) &&
+		curWinPlan.EqualFrame(ctx, nextWinPlan)) {
+		return false
+	}
+
+	// Whether the first window uses the unsettled columns in the next window.
+
+	// `select a, b, sum(bb) over (partition by a) as 'sum_bb' from (select a, b, max(b) over (partition by a) as 'bb' from t) as tt`
+	// The adjacent windows in the above sql statement cannot be merged.
+	// The reason is that the first one uses an unsettled column `bb` from the second one.
+	nextWindowCols := make(map[int64]struct{})
+	nextWindowChildrenExistedCols := make(map[int64]struct{})
+	for _, ngc := range nextGroupChildren {
+		for _, c := range ngc.Prop.Schema.Columns {
+			nextWindowChildrenExistedCols[c.UniqueID] = struct{}{}
+		}
+	}
+	for _, nc := range nextSchema.Columns {
+		nextWindowCols[nc.UniqueID] = struct{}{}
+	}
+	for _, funDesc := range curWinPlan.WindowFuncDescs {
+		for _, arg := range funDesc.Args {
+			cols := expression.ExtractColumns(arg)
+			for _, c := range cols {
+				_, ok1 := nextWindowCols[c.UniqueID]
+				_, ok2 := nextWindowChildrenExistedCols[c.UniqueID]
+				if ok1 && !ok2 {
+					return false
+				}
```

Comment:
It seems that the `!ok2` check is enough?

Reply:
> It seems that the `!ok2` check is enough?

@time-and-fate Maybe you are right. I considered the case `SELECT COUNT(1) OVER (...) FROM (SELECT COUNT(1) OVER (...) FROM T) AS TT` before, but I realize now that arg `1` can be ignored by `ExtractColumns()`. Maybe all columns in currentWindow arguments can be found in the nextWindowSchema.

Reply:
@time-and-fate Done. (https://github.com/pingcap/tidb/pull/24556/commits/4b0ec3fb502d4e1dbe042d8a379fe09937d354b8)